File: app/src/main/java/com/simplemobiletools/filemanager/dialogs/PropertiesDialog.java
package com.simplemobiletools.filemanager.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.text.format.DateFormat;
import android.view.View;
import android.widget.TextView;

import com.simplemobiletools.filemanager.Config;
import com.simplemobiletools.filemanager.R;
import com.simplemobiletools.filemanager.Utils;
import com.simplemobiletools.filemanager.models.FileDirItem;

import java.io.File;
import java.util.Calendar;
import java.util.Locale;

public class PropertiesDialog extends DialogFragment {
    private static FileDirItem mItem;
    private static int mFilesCnt;
    private static boolean mShowHidden;

    public static PropertiesDialog newInstance(FileDirItem item) {
        mItem = item;
        mFilesCnt = 0;
        return new PropertiesDialog();
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        mShowHidden = Config.newInstance(getContext()).getShowHidden();
        final int title = (mItem.getIsDirectory()) ? R.string.directory_properties : R.string.file_properties;

        final View infoView = getActivity().getLayoutInflater().inflate(R.layout.item_properties, null);
        ((TextView) infoView.findViewById(R.id.properties_name)).setText(mItem.getName());
        ((TextView) infoView.findViewById(R.id.properties_path)).setText(mItem.getPath());
        ((TextView) infoView.findViewById(R.id.properties_size)).setText(getItemSize());

        if (mItem.getIsDirectory()) {
            infoView.findViewById(R.id.properties_files_count_label).setVisibility(View.VISIBLE);
            infoView.findViewById(R.id.properties_files_count).setVisibility(View.VISIBLE);
            ((TextView) infoView.findViewById(R.id.properties_files_count)).setText(String.valueOf(mFilesCnt));
        }

        final File file = new File(mItem.getPath());
        ((TextView) infoView.findViewById(R.id.properties_last_modified)).setText(formatLastModified(file.lastModified()));

        final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle(getResources().getString(title));
        builder.setView(infoView);
        builder.setPositiveButton(R.string.ok, null);

        return builder.create();
    }

    private String getItemSize() {
        if (mItem.getIsDirectory()) {
            return Utils.formatSize(directorySize(new File(mItem.getPath())));
        }

        return Utils.getFormattedSize(mItem);
    }

    private String formatLastModified(long ts) {
        Calendar cal = Calendar.getInstance(Locale.ENGLISH);
        cal.setTimeInMillis(ts);
        return DateFormat.format("dd/MM/yyyy HH:mm", cal).toString();
    }

    private long directorySize(File dir) {
        if (dir.exists()) {
            long size = 0;
            File[] files = dir.listFiles();
            for (int i = 0; i < files.length; i++) {
                if (files[i].isDirectory()) {
                    size += directorySize(files[i]);
                } else {
                    size += files[i].length();
                    if ((!files[i].isHidden() && !dir.isHidden()) || mShowHidden)
                        mFilesCnt++;
                }
            }
            return size;
        }
        return 0;
    }
}

File: app/src/main/kotlin/com/simplemobiletools/filemanager/dialogs/PropertiesDialog.kt
package com.simplemobiletools.filemanager.dialogs

import android.content.Context
import android.content.res.Resources
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.ViewGroup
import com.simplemobiletools.filemanager.Config
import com.simplemobiletools.filemanager.R
import com.simplemobiletools.filemanager.extensions.*
import kotlinx.android.synthetic.main.item_properties.view.*
import kotlinx.android.synthetic.main.property_item.view.*
import java.io.File
import java.util.*

class PropertiesDialog() {
    lateinit var mContext: Context
    lateinit var mInflater: LayoutInflater
    lateinit var mPropertyView: ViewGroup
    lateinit var mResources: Resources

    private var mCountHiddenItems = false
    private var mFilesCnt = 0

    constructor(context: Context, path: String, countHiddenItems: Boolean = false) : this() {
        mContext = context
        mCountHiddenItems = countHiddenItems
        mInflater = LayoutInflater.from(context)
        mResources = mContext.resources
        mPropertyView = mInflater.inflate(R.layout.item_properties, null) as ViewGroup

        val file = File(path)
        addProperty(R.string.name, file.name)
        addProperty(R.string.path, file.parent)
        addProperty(R.string.size, getItemSize(file).formatSize())
        addProperty(R.string.last_modified, file.lastModified().formatLastModified())

        if (file.isDirectory) {
            addProperty(R.string.files_count, mFilesCnt.toString())
        } else if (file.isImage()) {
            addProperty(R.string.resolution, file.getImageResolution())
        } else if (file.isAudio()) {
            addProperty(R.string.duration, file.getDuration())
        } else if (file.isVideo()) {
            addProperty(R.string.duration, file.getDuration())
            addProperty(R.string.resolution, file.getVideoResolution())
        }

        AlertDialog.Builder(context)
                .setTitle(mResources.getString(R.string.properties))
                .setView(mPropertyView)
                .setPositiveButton(R.string.ok, null)
                .create()
                .show()
    }

    constructor(context: Context, paths: List<String>, countHiddenItems: Boolean = false) : this() {
        mContext = context
        mCountHiddenItems = countHiddenItems
        mInflater = LayoutInflater.from(context)
        mResources = mContext.resources
        mPropertyView = mInflater.inflate(R.layout.item_properties, null) as ViewGroup

        val files = ArrayList<File>(paths.size)
        paths.forEach { files.add(File(it)) }

        addProperty(R.string.path, files[0].parent)
        addProperty(R.string.size, getItemsSize(files).formatSize())
        addProperty(R.string.files_count, mFilesCnt.toString())

        AlertDialog.Builder(context)
                .setTitle(mResources.getString(R.string.properties))
                .setView(mPropertyView)
                .setPositiveButton(R.string.ok, null)
                .create()
                .show()
    }

    private fun addProperty(labelId: Int, value: String) {
        val view = mInflater.inflate(R.layout.property_item, mPropertyView, false)
        view.property_label.text = mResources.getString(labelId)
        view.property_value.text = value
        mPropertyView.properties_holder.addView(view)
    }

    private fun getItemsSize(files: ArrayList<File>): Long {
        var size = 0L
        files.forEach { size += getItemSize(it) }
        return size
    }

    private fun getItemSize(file: File): Long {
        if (file.isDirectory) {
            mCountHiddenItems = Config.newInstance(mContext).showHidden
            return getDirectorySize(File(file.path))
        }

        mFilesCnt++
        return file.length()
    }

    private fun getDirectorySize(dir: File): Long {
        var size = 0L
        if (dir.exists()) {
            val files = dir.listFiles()
            for (i in files.indices) {
                if (files[i].isDirectory) {
                    size += getDirectorySize(files[i])
                } else if (!files[i].isHidden && !dir.isHidden || mCountHiddenItems) {
                    mFilesCnt++
                    size += files[i].length()
                }
            }
        }
        return size
    }
}
